### PR TITLE
Use mashumaro to serialize/deserialize dataclass

### PR DESCRIPTION
**Pull request by hhcs9527**
# TL;DR

- Use mashumaro to serialize/deserialize dataclass
- the bigger task will speed up 1.5x process time in serialize/deserialize, with  benchmark workflow (provided below)
- performance testing [result sheet](https://docs.google.com/spreadsheets/d/1-e89sb_MZcg7i5nd6UDkYtbq2Kx6PU3I1t-KFCewfN8/edit#gid=300195406)

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description

1. Update the setup.py 
2. Extend the functionality of DataClassTransformer to support DataClassJSONMixin for better serialize/deserialize speed.
3. Complete code change for serialize/deserialize data class.
4. Compare old and new performance, benchmark workflow

```python
from flytekit import task, workflow, Resources
from typing import List, Tuple
from dataclasses_json import dataclass_json
from dataclasses import dataclass
from mashumaro.mixins.json import DataClassJSONMixin
from enum import Enum

@dataclass
class CurrencyPosition(DataClassJSONMixin):
    currency: str
    balance: float

@dataclass
class StockPosition(DataClassJSONMixin):
    ticker: str
    name: str
    balance: int

@dataclass
class OP(DataClassJSONMixin):
    currencies: List[CurrencyPosition]
    stocks: List[StockPosition]

@task(requests=Resources(mem="1Gi",cpu="1"),limits=Resources(mem="1Gi",cpu="1"),disable_deck=False)
def t1(obj: OP) -> OP:
    return obj

@task(requests=Resources(mem="1Gi",cpu="1"),limits=Resources(mem="1Gi",cpu="1"),disable_deck=False)
def t2(obj: OP) -> OP:
    return obj

@task(requests=Resources(mem="1Gi",cpu="1"),limits=Resources(mem="1Gi",cpu="1"),disable_deck=False)
def t3(obj: OP) -> OP:
    return obj

@workflow
def wf() -> OP:
    my_op = OP(
        currencies=[
            CurrencyPosition("USD", 238.67),
            CurrencyPosition("EUR", 361.84),
        ],
        stocks=[
            StockPosition("AAPL", "Apple", 10),
            StockPosition("AMZN", "Amazon", 10),
        ]
    )

    res = t1(obj=my_op)
    res = t2(obj=res)
    res = t3(obj=res)

    return res
```

### Performance Compare 

- We compare three items by running the three items 3 times for each then find the avergae time of to_literal and to_python val
- the result shows, we reach **1.5x faster** than previous version of DataClassTransformer with **DataclassJSONMixin as input**

1. gcp_dataclass_json_new, test the dataclass_json version of workflow (by simply change the input) with new version of DataClassTransformer
2. gcp_dataclassJSONMixin, test the DataclassJSONMixin version of workflow (by simply change the input) with new version of DataClassTransformer
3. gcp_dataclass_json_old, test the dataclass_json version of workflow (by simply change the input) with older version of DataClassTransformer

**Compare to_literal**
image

**Compare to_python val**
image

## Tracking Issue
[https://github.com/flyteorg/flyte/issues/3364](https://github.com/flyteorg/flyte/issues/3364)

### CODE REVIEW

**Review by TaiJuWu — APPROVED** [review]

File: flytekit/core/type_engine.py [hunk 1]
```diff
@@ -353,7 +354,7 @@ def to_literal(self, ctx: FlyteContext, python_val: T, python_type: Type[T], exp
             )
         self._serialize_flyte_type(python_val, python_type)
         return Literal(
-            scalar=Scalar(generic=_json_format.Parse(cast(DataClassJsonMixin, python_val).to_json(), _struct.Struct()))
+            scalar=Scalar(generic=_json_format.Parse(cast(DataClassJSONMixin, python_val).to_json(), _struct.Struct()))
```
Comment (pingsutw) [inline flytekit/core/type_engine.py:357]: Instead of using `cast`, we should apply mixins to a class instance after creation. check https://stackoverflow.com/questions/8544983/dynamically-mixin-a-base-class-to-an-instance-in-python.

`cast` does nothing at runtime.

File: requirements.txt [hunk 2]
```diff
@@ -0,0 +1,376 @@
+#
```
Comment (pingsutw) [inline requirements.txt:1]: Actually, we don't need this file.

**Comment by hhcs9527** [comment]
@pingsutw , just finish the changes and the relevant tests, please take a look.

File: flytekit/core/type_engine.py [hunk 3]
```diff
@@ -347,14 +355,21 @@ def to_literal(self, ctx: FlyteContext, python_val: T, python_type: Type[T], exp
                 f"{type(python_val)} is not of type @dataclass, only Dataclasses are supported for "
                 f"user defined datatypes in Flytekit"
             )
-        if not issubclass(type(python_val), DataClassJsonMixin):
+        if not issubclass(type(python_val), DataClassJsonMixin) and not issubclass(
+            type(python_val), DataClassJSONMixin
+        ):
```
Comment (pingsutw) [inline flytekit/core/type_engine.py:360]:
```suggestion
        if not isinstance(python_val, (DataClassJsonMixin, DataClassJSONMixin)):
```

File: flytekit/core/type_engine.py [hunk 4]
```diff
@@ -598,13 +613,20 @@ def to_python_value(self, ctx: FlyteContext, lv: Literal, expected_python_type:
...
+                f"Dataclass {expected_python_type} should be decorated with @dataclass_json or mixin with DataClassJSONMixin to be "
                 f"serialized correctly"
             )
         json_str = _json_format.MessageToJson(lv.scalar.generic)
-        dc = cast(DataClassJsonMixin, expected_python_type).from_json(json_str)
+
+        if issubclass(expected_python_type, DataClassJsonMixin):
+            dc = cast(DataClassJsonMixin, expected_python_type).from_json(json_str)  # type: ignore
+        else:
+            dc = cast(DataClassJSONMixin, expected_python_type).from_json(json_str)  # type: ignore
```
Comment (pingsutw) [inline flytekit/core/type_engine.py:628]:
```suggestion
        dc = cast(DataClassJSONMixin, expected_python_type).from_json(json_str)  # type: ignore
```

File: flytekit/types/file/file.py [hunk 5]
```diff
@@ -25,9 +26,8 @@ def noop():
 T = typing.TypeVar("T")
 
 
-@dataclass_json
 @dataclass
-class FlyteFile(os.PathLike, typing.Generic[T]):
+class FlyteFile(os.PathLike, typing.Generic[T], DataClassJSONMixin):
     path: typing.Union[str, os.PathLike] = field(
```
Comment (pingsutw) [inline flytekit/types/file/file.py:31]: We still use `marshmallow` here. Are we able to use `mashumaro`.field or any equivalent

File: flytekit/core/type_engine.py (hunk 4 above)
Comment (hhcs9527) [reply flytekit/core/type_engine.py:628]: Hi @pingsutw, some problems to ask.
we keep 
 dc = cast(DataClassJSONMixin, expected_python_type).from_json(json_str)  with mashumaro,
but we use 
json_str = cast(DataClassJsonMixin, python_val).to_json()  # type: ignore with dataclass json

Should we be consistent through the changes?

File: flytekit/types/file/file.py (hunk 5 above)
Comment (hhcs9527) [reply flytekit/types/file/file.py:31]: Hi @pingsutw,
Can you be more specific about this? the local will pass all test.

**Comment by hhcs9527** [comment]
@pingsutw , looks like all the test has been pass, please take a looks at this pr

File: flytekit/core/type_engine.py [hunk 6]
```diff
@@ -614,10 +624,13 @@ def to_python_value(self, ctx: FlyteContext, lv: Literal, expected_python_type:
...
     def guess_python_type(self, literal_type: LiteralType) -> Type[T]:  # type: ignore
         if literal_type.simple == SimpleType.STRUCT:
-            if literal_type.metadata is not None and DEFINITIONS in literal_type.metadata:
-                schema_name = literal_type.metadata["$ref"].split("/")[-1]
-                return convert_json_schema_to_python_class(literal_type.metadata[DEFINITIONS], schema_name)
-
+            if literal_type.metadata is not None:
+                if DEFINITIONS in literal_type.metadata:
+                    schema_name = literal_type.metadata["$ref"].split("/")[-1]
+                    return convert_json_schema_to_python_class(literal_type.metadata[DEFINITIONS], schema_name)
```
Comment (pingsutw) [inline flytekit/core/type_engine.py:630]: could we create two functions for this? 
1. convert_mashumaro_json_schema_to_python_class
2. convert_marshmallow_json_schema_to_python_class

**Comment by hhcs9527** [comment]
Hi @pingsutw , just finished the change of the code by splitting the functions, please approve the testing and take a look.

**Comment by hhcs9527** [comment]
Hi @pingsutw , please take a look at the pr, thanks.

**Comment by hhcs9527** [comment]
Hi @pingsutw, just finish the changes and the relevant tests, please approve the test.

File: flytekit/core/type_engine.py [hunk 7]
```diff
@@ -1575,13 +1626,34 @@ def convert_json_schema_to_python_class(schema: Dict[str, Any], schema_name: str
...
+    Generate a model class based on the provided JSON Schema
+    :param schema: dict representing valid JSON schema
+    :param schema_name: dataclass name of return type
+    """
+
+    attribute_list = generate_attribute_list_from_dataclass_json(schema, schema_name)
+    return dataclass_json(dataclasses.make_dataclass(schema_name, attribute_list))
+
+
+def convert_mashumaro_json_schema_to_python_class(schema: dict, schema_name: typing.Any, is_dataclass_json_mixin: bool = False) -> Type[dataclasses.dataclass()]:  # type: ignore
```
Comment (pingsutw) [inline flytekit/core/type_engine.py:1644]:
```suggestion
def convert_mashumaro_json_schema_to_python_class(schema: dict, schema_name: typing.Any) -> Type[dataclasses.dataclass()]:  # type: ignore
```

File: flytekit/core/type_engine.py [hunk 8]
```diff
@@ -1575,13 +1626,34 @@ def convert_json_schema_to_python_class(schema: Dict[str, Any], schema_name: str
                 attribute_list.append((property_key, Dict[str, _get_element_type(property_val)]))  # type: ignore[misc,index]
         # Handle int, float, bool or str
         else:
-            attribute_list.append((property_key, _get_element_type(property_val)))
+            attribute_list.append([property_key, _get_element_type(property_val)])  # type: ignore
+    return attribute_list
+
+
+def convert_marshmallow_json_schema_to_python_class(schema: dict, schema_name: typing.Any, is_dataclass_json_mixin: bool = False) -> Type[dataclasses.dataclass()]:  # type: ignore
```
Comment (pingsutw) [inline flytekit/core/type_engine.py:1633]:
```suggestion
def convert_marshmallow_json_schema_to_python_class(schema: dict, schema_name: typing.Any) -> Type[dataclasses.dataclass()]:  # type: ignore
```

File: tests/flytekit/unit/core/test_type_engine.py [hunk 9]
```diff
@@ -700,6 +893,105 @@ class TestFileStruct(DataClassJsonMixin):
...
+        # assert lv.scalar.generic["c"]["path"] == remote_path
+        # assert lv.scalar.generic["d"]["path"] == remote_path
+        # assert lv.scalar.generic["e"]["path"]== remote_path
+        # assert lv.scalar.generic["e_prime"].values[0].WhichOneof("kind") == "null_value"
+        # assert lv.scalar.generic["f"]["a"]["path"] == remote_path
+        # assert lv.scalar.generic["g"]["a"]["path"] == remote_path
+        # assert lv.scalar.generic["g_prime"]["a"] is None
+        # assert lv.scalar.generic["h"]["path"] == remote_path
+        # assert lv.scalar.generic["h_prime"] is None
+        # assert lv.scalar.generic["i"]["a"] == 42
```
Comment (pingsutw) [inline tests/flytekit/unit/core/test_type_engine.py:959]: delete it?

Comment (hhcs9527) [reply tests/flytekit/unit/core/test_type_engine.py:959]: sure, thanks for reminding me.

**Comment by hhcs9527** [comment]
Hi @pingsutw, please approve the CI test.

File: flytekit/core/type_engine.py [hunk 10]
```diff
@@ -376,15 +384,21 @@ def to_literal(self, ctx: FlyteContext, python_val: T, python_type: Type[T], exp
...
             )
         self._serialize_flyte_type(python_val, python_type)
-        return Literal(
-            scalar=Scalar(generic=_json_format.Parse(cast(DataClassJsonMixin, python_val).to_json(), _struct.Struct()))
-        )
+
+        if issubclass(type(python_val), DataClassJsonMixin):
+            json_str = cast(DataClassJsonMixin, python_val).to_json()  # type: ignore
+        else:
+            json_str = cast(DataClassJSONMixin, python_val).to_json()  # type: ignore
```
Comment (pingsutw) [inline flytekit/core/type_engine.py:399]:
```suggestion
        json_str = python_val.to_json()  # type: ignore
```

File: dev-requirements.in [hunk 11]
```diff
@@ -11,6 +11,7 @@ pre-commit
 codespell
 google-cloud-bigquery
 google-cloud-bigquery-storage
+snowflake-connector-python
```
Comment (pingsutw) [inline dev-requirements.in:14]: remove it ?

Comment (hhcs9527) [reply dev-requirements.in:14]: sure

**Comment by hhcs9527** [comment]
Hi @pingsutw, just finished the changes, Please take a look.

**Review by pingsutw — APPROVED** [review]
